hummingbot/connector/exchange/k2/k2_api_order_book_data_source.py

```diff
@@ -0,0 +1,233 @@
+                            data = ujson.loads(response["data"])
+                            trading_pair = k2_utils.convert_from_exchange_trading_pair(data["pair"])
+                            message: OrderBookMessage = K2OrderBook.diff_message_from_exchange(
+                                msg=data,
+                                timestamp=timestamp,
+                                metadata={"trading_pair": trading_pair})
+                        else:
+                            # Ignores all other messages
+                            continue
+                        output.put_nowait(message)
+                except asyncio.CancelledError:
+                    raise
+                except Exception:
+                    self.logger().network(
+                        "Unexpected error with WebSocket connection.",
+                        exc_info=True,
+                        app_warning_msg="Unexpected error with WebSocket connection. Retrying in 30 seconds. "
+                                        "Check network connection."
+                    )
+                    await asyncio.sleep(30.0)
+                finally:
+                    await ws.close()
+
+    async def listen_for_order_book_snapshots(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        """
+        Listen for orderbook snapshots by fetching orderbook using the REST API.
+        """
+        while True:
+            try:
+                for trading_pair in self._trading_pairs:
+                    try:
+                        snapshot: Dict[str, any] = await self.get_order_book_data(trading_pair)
+                        snapshot_timestamp: int = int(time.time() * 1e3)
+                        snapshot_msg: OrderBookMessage = K2OrderBook.snapshot_message_from_exchange(
+                            snapshot,
+                            snapshot_timestamp,
+                            metadata={"trading_pair": trading_pair}
+                        )
+                        output.put_nowait(snapshot_msg)
+                        self.logger().debug(f"Saved order book snapshot for {trading_pair}")
+                        # Be careful not to go above API rate limits.
+                        await asyncio.sleep(5.0)
+                    except asyncio.CancelledError:
+                        raise
+                    except Exception:
+                        self.logger().network(
+                            "Unexpected error occured retrieving Order Book Data using REST API. Retying in 5 seconds",
+                            exc_info=True
+                        )
+                        await asyncio.sleep(5.0)
+                this_hour: pd.Timestamp = pd.Timestamp.utcnow().replace(minute=0, second=0, microsecond=0)
+                next_hour: pd.Timestamp = this_hour + pd.Timedelta(hours=1)
+                delta: float = next_hour.timestamp() - time.time()
+                await asyncio.sleep(delta)
```

Comment:
1 snapshot per hour? Isn't that too much? That's an exchange's limitation?

Reply:
Reduce snapshot interval to once every 5mins